File: 백준/Gold/20058. 마법사 상어와 파이어스톰/마법사 상어와 파이어스톰.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int q = Integer.parseInt(st.nextToken());

        int[][] arr = new int[1 << n][1 << n];

        for (int i = 0; i < arr.length; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < arr.length; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < q; i++) {
            int l = Integer.parseInt(st.nextToken());

            arr = rotate(arr, l);
            decrease(arr);
        }

        System.out.println(getSum(arr));
        System.out.println(getMaxSize(arr));
    }

    private static int[][] rotate(int[][] arr, int l) {

        int n = arr.length;

        int[][] newArr = new int[n][n];
        int size = 1 << l;

        for (int i = 0; i < n; i += size) {
            for (int j = 0; j < n; j += size) {

                for (int x = 0; x < size; x++) {
                    for (int y = 0; y < size; y++) {
                        newArr[i + y][j + size - 1 - x] = arr[i + x][j + y];
                    }
                }
            }
        }

        return newArr;
    }

    private static void decrease(int[][] arr) {

        int n = arr.length;
        boolean[][] mark = new boolean[n][n];

        for (int x = 0; x < n; x++) {
            for (int y = 0; y < n; y++) {

                int count = 0;

                for (int d = 0; d < 4; d++) {
                    int nx = x + dx[d];
                    int ny = y + dy[d];

                    if (nx < 0 || ny < 0 || nx >= n || ny >= n || arr[nx][ny] <= 0) {
                        continue;
                    }
                    count++;
                }

                if (count < 3) {
                    mark[x][y] = true;
                }
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (mark[i][j]) {
                    arr[i][j]--;
                }
            }
        }
    }

    private static int getSum(int[][] arr) {
        int sum = 0;
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr.length; j++) {
                sum += Math.max(arr[i][j], 0);
            }
        }
        return sum;
    }

    private static int getMaxSize(int[][] arr) {

        int n = arr.length;
        boolean[][] visit = new boolean[n][n];

        int maxSize = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (arr[i][j] > 0 && !visit[i][j]) {
                    maxSize = Math.max(maxSize, bfs(i, j, arr, visit));
                }
            }
        }

        return maxSize;
    }

    private static int bfs(int i, int j, int[][] arr, boolean[][] visit) {

        int size = 1;
        visit[i][j] = true;
        int n = arr.length;

        Queue<Integer> qu = new ArrayDeque<>();
        qu.offer(i);
        qu.offer(j);

        while (!qu.isEmpty()) {
            int x = qu.poll();
            int y = qu.poll();

            for (int d = 0; d < 4; d++) {
                int nx = x + dx[d];
                int ny = y + dy[d];

                if (nx < 0 || ny < 0 || nx >= n || ny >= n || visit[nx][ny] || arr[nx][ny] <= 0) {
                    continue;
                }
                size++;
                visit[nx][ny] = true;
                qu.offer(nx);
                qu.offer(ny);
            }
        }

        return size;
    }
}

File: 백준/Gold/20058. 마법사 상어와 파이어스톰/마법사 상어와 파이어스톰.py
from collections import deque

dx = (-1, 1, 0, 0)
dy = (0, 0, -1, 1)


def rotate(arr, l):
    n = len(arr)
    new_arr = [[0] * n for _ in range(n)]
    size = 1 << l

    for i in range(0, n, size):
        for j in range(0, n, size):
            for x in range(size):
                for y in range(size):
                    new_arr[i + y][j + size - x - 1] = arr[i + x][j + y]

    return new_arr


def decrease(arr):
    n = len(arr)
    mark = [[False] * n for _ in range(n)]

    for x in range(n):
        for y in range(n):

            count = 0

            for d in range(4):
                nx = x + dx[d]
                ny = y + dy[d]

                if nx < 0 or ny < 0 or nx >= n or ny >= n or arr[nx][ny] <= 0:
                    continue

                count += 1

            if count < 3:
                mark[x][y] = True

    for i in range(n):
        for j in range(n):
            if mark[i][j]:
                arr[i][j] -= 1


def get_sum(arr):
    return sum(max(0, value) for row in arr for value in row)


def get_max_size(arr):
    n = len(arr)
    visit = [[False] * n for _ in range(n)]
    max_size = 0

    for i in range(n):
        for j in range(n):
            if arr[i][j] > 0 and not visit[i][j]:
                max_size = max(max_size, bfs(i, j, arr, visit))

    return max_size


def bfs(i, j, arr, visit):
    size = 1
    n = len(arr)
    visit[i][j] = True

    qu = deque([(i, j)])

    while qu:
        x, y = qu.popleft()

        for d in range(4):
            nx = x + dx[d]
            ny = y + dy[d]

            if nx < 0 or ny < 0 or nx >= n or ny >= n or visit[nx][ny] or arr[nx][ny] <= 0:
                continue

            size += 1
            visit[nx][ny] = True
            qu.append((nx, ny))

    return size


n, q = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(1 << n)]
queries = list(map(int, input().split()))

for query in queries:
    arr = rotate(arr, query)
    decrease(arr)

print(get_sum(arr))
print(get_max_size(arr))